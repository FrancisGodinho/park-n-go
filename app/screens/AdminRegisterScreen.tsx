import {
  StyleSheet,
  Text,
  SafeAreaView,
  TouchableWithoutFeedback,
  Keyboard,
  TextInput,
  View,
  TouchableOpacity,
} from "react-native";
import React, { useRef, useState } from "react";
import {
  createUserWithEmailAndPassword,
  getAuth,
  updateProfile,
  User,
} from "firebase/auth";
import { doc, setDoc, Timestamp } from "@firebase/firestore";
import { useFormik } from "formik";
import * as Yup from "yup";
import Headers from "../constants/Headers";
import { auth, db } from "../utils/Firebase";
import CustomTextInput from "../components/CustomTextInput";
import Colors from "../constants/Colors";
import CustomButton from "../components/CustomButton";
import AuthRedirect from "../components/AuthRedirect";
import { NativeStackScreenProps } from "@react-navigation/native-stack";
import { StackParamList } from "../types";

const RegisterSchema = Yup.object().shape({
  name: Yup.string().required("Required"),
  company: Yup.string().required("Required"),
  email: Yup.string().email("Invalid email").required("Required"),
  password: Yup.string()
    .min(6, "Must be 6 characters or more")
    .required("Required"),
  confirmPassword: Yup.string()
    .oneOf([Yup.ref("password"), null], "Passwords must match")
    .required("Required"),
});

const errorCodes: object = {
  "auth/invalid-email": "Invalid email",
};

const AdminRegisterScreen = ({
  navigation,
}: NativeStackScreenProps<StackParamList, "AdminRegister">) => {
  const [firebaseError, setFirebaseError] = useState({ code: "", message: "" });

  const refCompany = useRef<TextInput>();
  const refEmail = useRef<TextInput>();
  const refPassword = useRef<TextInput>();
  const refConfirmPassword = useRef<TextInput>();

  const formik = useFormik({
    initialValues: {
      name: "",
      company: "",
      email: "",
      password: "",
      confirmPassword: "",
    },
    validationSchema: RegisterSchema,
    onSubmit: (values, { resetForm }) => {
      signUp(values);
      resetForm();
    },
  });

  // Sign up with email
  const signUp = async ({
    name,
    email,
    password,
  }: {
    name: string;
    email: string;
    password: string;
  }) => {
    createUserWithEmailAndPassword(auth, email, password)
      .then((cred) => {
        updateProfile(getAuth().currentUser as User, { displayName: name });
        setDoc(doc(db, "users", cred.user.uid), {
          dateCreated: Timestamp.fromDate(new Date()),
          displayName: name,
          streak: 0,
          isAdmin: true,
        });
      })
      .catch((err) => setFirebaseError(err));
  };

  return (
    <TouchableWithoutFeedback onPress={Keyboard.dismiss} accessible={false}>
      <SafeAreaView style={styles.container}>
        <Text style={[Headers.h1, styles.h1]}>Create Admin Account</Text>
        <CustomTextInput
          formik={formik}
          icon="user"
          name="name"
          placeholder="Name"
          value={formik.values.name}
          onSubmitEditing={() => refCompany.current?.focus()}
        />
        {formik.touched.name && formik.errors.name ? (
          <Text style={[Headers.p, styles.error]}>{formik.errors.name}</Text>
        ) : null}
        <CustomTextInput
          formik={formik}
          icon="building"
          name="company"
          placeholder="Company"
          refValue={refCompany}
          value={formik.values.company}
          onSubmitEditing={() => refEmail.current?.focus()}
        />
        {formik.touched.company && formik.errors.company ? (
          <Text style={[Headers.p, styles.error]}>{formik.errors.name}</Text>
        ) : null}
        <CustomTextInput
          formik={formik}
          icon="envelope"
          keyboardType="email-address"
          name="email"
          placeholder="E-mail"
          refValue={refEmail}
          value={formik.values.email}
          onSubmitEditing={() => refPassword.current?.focus()}
        />
        {formik.touched.email && formik.errors.email ? (
          <Text style={[Headers.p, styles.error]}>{formik.errors.email}</Text>
        ) : null}
        <CustomTextInput
          formik={formik}
          icon="lock"
          isPass
          name="password"
          placeholder="Password"
          refValue={refPassword}
          value={formik.values.password}
          onSubmitEditing={() => refConfirmPassword.current?.focus()}
        />
        {formik.touched.password && formik.errors.password ? (
          <Text style={[Headers.p, styles.error]}>
            {formik.errors.password}
          </Text>
        ) : null}
        <CustomTextInput
          formik={formik}
          icon="lock"
          isPass
          name="confirmPassword"
          placeholder="Confirm Password"
          refValue={refConfirmPassword}
          value={formik.values.confirmPassword}
          onSubmitEditing={formik.handleSubmit}
        />
        {formik.touched.confirmPassword && formik.errors.confirmPassword ? (
          <Text style={[Headers.p, styles.error]}>
            {formik.errors.confirmPassword}
          </Text>
        ) : null}
        <View style={styles.bottom}>
          <View style={styles.admin}>
            <TouchableOpacity
              onPress={() => navigation.navigate("UserRegister")}
            >
              <Text
                style={[
                  Headers.p,
                  {
                    color: Colors.primary,
                    fontSize: 15,
                  },
                ]}
              >
                Register
              </Text>
            </TouchableOpacity>
            <Text
              style={[
                Headers.p,
                {
                  fontSize: 15,
                },
              ]}
            >
              {" "}
              as a user
            </Text>
          </View>

          <CustomButton
            disabled={
              !formik.values.name ||
              !formik.values.email ||
              !formik.values.password ||
              !formik.values.confirmPassword
            }
            text="Register"
            onPress={formik.handleSubmit}
          />
        </View>
        <Text style={[Headers.p, styles.error]}>
          {Object.keys(errorCodes).includes(firebaseError.code)
            ? errorCodes[firebaseError.code as keyof object]
            : firebaseError.message}
        </Text>
        <AuthRedirect toScreen="Login" />
      </SafeAreaView>
    </TouchableWithoutFeedback>
  );
};

export default AdminRegisterScreen;

const styles = StyleSheet.create({
  container: { flex: 1, marginHorizontal: 10 },
  h1: { marginTop: 20, marginBottom: 10 },
  bottom: { marginTop: "auto" },
  admin: { flexDirection: "row", justifyContent: "center", marginBottom: 30 },
  error: {
    color: "red",
    marginLeft: 10,
    fontSize: 15,
  },
});
