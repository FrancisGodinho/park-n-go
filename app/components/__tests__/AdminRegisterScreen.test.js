import React from "react";
import renderer from "react-test-renderer";
import { cleanup } from "@testing-library/react-native";
import { fireEvent, render } from "@testing-library/react";
import { NavigationContainer } from "@react-navigation/native";
import AdminRegisterScreen from "../../screens/AdminRegisterScreen";

beforeAll((done) => {
  done();
});

afterAll((done) => {
  done();
});

let findTextElement = function (tree, element) {
  console.warn(tree);
  return true;
};

describe("<AdminRegisterScreen />", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <NavigationContainer>
          <AdminRegisterScreen />
        </NavigationContainer>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
