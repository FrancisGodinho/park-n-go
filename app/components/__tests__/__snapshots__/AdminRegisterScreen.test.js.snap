// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminRegisterScreen /> renders correctly 1`] = `
<RCTSafeAreaView
  accessible={false}
  emulateUnlessSupported={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flex": 1,
      "marginHorizontal": 10,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#C0C0C0",
          "fontFamily": "Avenir-Black",
          "fontSize": 25,
          "fontWeight": "900",
          "textAlign": "center",
        },
        Object {
          "marginBottom": 10,
          "marginTop": 20,
        },
      ]
    }
  >
    Create Admin Account
  </Text>
  <View
    onLayout={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#404040",
        "borderRadius": 5,
        "flexDirection": "row",
        "marginVertical": 10,
        "paddingVertical": 10,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "marginLeft": 10,
          "width": 30,
        }
      }
    >
      <Text />
    </View>
    <TextInput
      allowFontScaling={true}
      autoCorrect={false}
      editable={true}
      enablesReturnKeyAutomatically={true}
      keyboardType="default"
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Name"
      placeholderTextColor="#A0A0A0"
      rejectResponderTermination={true}
      returnKeyType="next"
      style={
        Array [
          Object {
            "color": "#C0C0C0",
            "fontFamily": "Avenir-Black",
            "fontSize": 17,
            "fontWeight": "400",
          },
          Object {
            "flex": 1,
            "paddingHorizontal": 10,
          },
          undefined,
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    onLayout={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#404040",
        "borderRadius": 5,
        "flexDirection": "row",
        "marginVertical": 10,
        "paddingVertical": 10,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "marginLeft": 10,
          "width": 30,
        }
      }
    >
      <Text />
    </View>
    <TextInput
      allowFontScaling={true}
      autoCorrect={false}
      editable={true}
      enablesReturnKeyAutomatically={true}
      keyboardType="default"
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Company"
      placeholderTextColor="#A0A0A0"
      rejectResponderTermination={true}
      returnKeyType="next"
      style={
        Array [
          Object {
            "color": "#C0C0C0",
            "fontFamily": "Avenir-Black",
            "fontSize": 17,
            "fontWeight": "400",
          },
          Object {
            "flex": 1,
            "paddingHorizontal": 10,
          },
          undefined,
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    onLayout={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#404040",
        "borderRadius": 5,
        "flexDirection": "row",
        "marginVertical": 10,
        "paddingVertical": 10,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "marginLeft": 10,
          "width": 30,
        }
      }
    >
      <Text />
    </View>
    <TextInput
      allowFontScaling={true}
      autoCorrect={false}
      editable={true}
      enablesReturnKeyAutomatically={true}
      keyboardType="email-address"
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="E-mail"
      placeholderTextColor="#A0A0A0"
      rejectResponderTermination={true}
      returnKeyType="next"
      style={
        Array [
          Object {
            "color": "#C0C0C0",
            "fontFamily": "Avenir-Black",
            "fontSize": 17,
            "fontWeight": "400",
          },
          Object {
            "flex": 1,
            "paddingHorizontal": 10,
          },
          undefined,
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    onLayout={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#404040",
        "borderRadius": 5,
        "flexDirection": "row",
        "marginVertical": 10,
        "paddingVertical": 10,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "marginLeft": 10,
          "width": 30,
        }
      }
    >
      <Text />
    </View>
    <TextInput
      allowFontScaling={true}
      autoCorrect={false}
      editable={true}
      enablesReturnKeyAutomatically={true}
      keyboardType="default"
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Password"
      placeholderTextColor="#A0A0A0"
      rejectResponderTermination={true}
      returnKeyType="next"
      secureTextEntry={true}
      style={
        Array [
          Object {
            "color": "#C0C0C0",
            "fontFamily": "Avenir-Black",
            "fontSize": 17,
            "fontWeight": "400",
          },
          Object {
            "flex": 1,
            "paddingHorizontal": 10,
          },
          undefined,
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginRight": 10,
          "opacity": 1,
        }
      }
    >
      <Text />
    </View>
  </View>
  <View
    onLayout={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#404040",
        "borderRadius": 5,
        "flexDirection": "row",
        "marginVertical": 10,
        "paddingVertical": 10,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "marginLeft": 10,
          "width": 30,
        }
      }
    >
      <Text />
    </View>
    <TextInput
      allowFontScaling={true}
      autoCorrect={false}
      editable={true}
      enablesReturnKeyAutomatically={true}
      keyboardType="default"
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Confirm Password"
      placeholderTextColor="#A0A0A0"
      rejectResponderTermination={true}
      returnKeyType="next"
      secureTextEntry={true}
      style={
        Array [
          Object {
            "color": "#C0C0C0",
            "fontFamily": "Avenir-Black",
            "fontSize": 17,
            "fontWeight": "400",
          },
          Object {
            "flex": 1,
            "paddingHorizontal": 10,
          },
          undefined,
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginRight": 10,
          "opacity": 1,
        }
      }
    >
      <Text />
    </View>
  </View>
  <View
    style={
      Object {
        "marginTop": "auto",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "center",
          "marginBottom": 30,
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#C0C0C0",
                "fontFamily": "Avenir-Black",
                "fontSize": 17,
                "fontWeight": "400",
              },
              Object {
                "color": "#66A182",
                "fontSize": 15,
              },
            ]
          }
        >
          Register
        </Text>
      </View>
      <Text
        style={
          Array [
            Object {
              "color": "#C0C0C0",
              "fontFamily": "Avenir-Black",
              "fontSize": 17,
              "fontWeight": "400",
            },
            Object {
              "fontSize": 15,
            },
          ]
        }
      >
         
        as a user
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#404040",
          "borderRadius": 5,
          "marginBottom": 10,
          "marginTop": "auto",
          "opacity": 1,
          "paddingVertical": 10,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#C0C0C0",
              "fontFamily": "Avenir-Black",
              "fontSize": 17,
              "fontWeight": "400",
            },
            Object {
              "color": "#E0E0E0",
              "textAlign": "center",
            },
            Object {
              "color": "#C0C0C0",
            },
          ]
        }
      >
        Register
      </Text>
    </View>
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#C0C0C0",
          "fontFamily": "Avenir-Black",
          "fontSize": 17,
          "fontWeight": "400",
        },
        Object {
          "color": "red",
          "fontSize": 15,
          "marginLeft": 10,
        },
      ]
    }
  >
    
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 20,
        "marginTop": 10,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#C0C0C0",
            "fontFamily": "Avenir-Black",
            "fontSize": 17,
            "fontWeight": "400",
          },
          Object {
            "fontSize": 15,
          },
        ]
      }
    >
      Already
       have an account?
       
    </Text>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#C0C0C0",
              "fontFamily": "Avenir-Black",
              "fontSize": 17,
              "fontWeight": "400",
            },
            Object {
              "color": "#66A182",
              "fontSize": 15,
            },
          ]
        }
      >
        Login
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
